pteranodon/server_detector.py

```diff
@@ -0,0 +1,298 @@
+        :param conn_timeout: Timeout for heartbeat message wait
+        :type conn_timeout: float
+        :return: None if no server at path, a dictionary of parameters otherwise
+        :rtype: Optional[Dict[str, Any]]
+        """
+        if serial_dev is None and port is None and is_tcp is None:
+            return None
+
+        conn: Optional[mavutil.mavfile] = None
+
+        if serial_dev is not None:
+            try:
+                conn = mavutil.mavlink_connection(device=serial_dev, baud=115200)
+            except serialutil.SerialException:
+                return None
+
+        if port is not None:
+            prefix = "tcp" if is_tcp else "udp"
+            addr = f"{prefix}:{self._server_addr}:{port}"
+            conn = mavutil.mavlink_connection(addr)
+
+        # Invalid parameters
+        if conn is None:
+            return None
+
+        msg = conn.recv_match(type="HEARTBEAT", blocking=True, timeout=conn_timeout)
+
+        conn.close()
+
+        # Msg is None on timeout
+        if msg is None:
+            return None
+
+        return {
+            "serial_dev": serial_dev,
+            "port": port,
+            "is_tcp": is_tcp,
+        }
+
+    # For the given dictionary, get the corresponding string server path
+    def _get_server_value(self, data_dict: Optional[Dict[str, Any]]) -> Optional[str]:
+        if data_dict is None:
+            return None
+
+        serial_dev: Optional[str] = data_dict["serial_dev"]
+        port: Optional[int] = data_dict["port"]
+
+        if serial_dev is not None:
+            return f"serial://{serial_dev}:{self._serial_baud_rate}"
+
+        addr = self._server_addr if self._server_addr != "127.0.0.1" else ""
+
+        if data_dict["is_tcp"]:
+            return f"tcp://{addr}:{port}"
+
+        return f"udp://{addr}:{port}"
+
+    async def _get_mavsdk_servers_data(
+        self, test_cached: bool
+    ) -> List[List[Optional[Dict]]]:
+        serial_devs = ServerDetector._fetch_possible_serial_devs()
+
+        tcp_ports = await self.fetch_open_proto_ports(
+            socket.SOCK_STREAM, fetch_cached=test_cached
+        )
+        udp_ports = await self.fetch_open_proto_ports(
+            socket.SOCK_DGRAM, fetch_cached=test_cached
+        )
+
+        serial_devs_future = asyncio.gather(
+            *[self._test_server(serial_dev=dev) for dev in serial_devs]
+        )
+
+        tcp_ports_future = asyncio.gather(
+            *[self._test_server(port=port, is_tcp=True) for port in tcp_ports]
+        )
+        udp_ports_future = asyncio.gather(
+            *[self._test_server(port=port, is_tcp=False) for port in udp_ports]
+        )
+
+        all_data = await asyncio.gather(
+            serial_devs_future, tcp_ports_future, udp_ports_future
+        )
+
+        # This is the correct type. `gather` states that it returns lists,
+        #   it returns lists in the code, but mypy believes it returns a tuple.
+        return all_data  # type: ignore
```

Comment:
Could cast all_data into a list with unpacking
```suggestion
        return [*all_data]
```
This could review any type errors

Reply:
This works, but I am changing it locally. Thank you